File: sdk/go/provider/throwawaymail.go
package provider

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/url"
	"strings"

	http "github.com/bogdanfinn/fhttp"
)

const throwawaymailBase = "https://throwawaymail.app/api"

type throwawaymailMailboxResponse struct {
	MailboxID string `json:"mailbox_id"`
	Address   string `json:"address"`
	ExpiresAt string `json:"expires_at"`
	CreatedAt string `json:"created_at"`
}

type throwawaymailMessageSummary struct {
	MessageID   string  `json:"message_id"`
	FromAddress string  `json:"from_address"`
	FromName    *string `json:"from_name"`
	Subject     string  `json:"subject"`
	ReceivedAt  string  `json:"received_at"`
	Size        int     `json:"size"`
	Read        bool    `json:"read"`
}

type throwawaymailMessageDetail struct {
	throwawaymailMessageSummary
	ToAddresses  []string `json:"to_addresses"`
	CCAddresses  []string `json:"cc_addresses"`
	BCCAddresses []string `json:"bcc_addresses"`
	ReplyTo      *string  `json:"reply_to"`
	Text         *string  `json:"text"`
	HTML         *string  `json:"html"`
}

func throwawaymailJSON(method, u string, body []byte, out any) error {
	req, err := http.NewRequest(method, u, bytes.NewReader(body))
	if err != nil {
		return err
	}
	req.Header.Set("Accept", "application/json")
	if body != nil {
		req.Header.Set("Content-Type", "application/json")
	}
	resp, err := HTTPClient().Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	raw, err := io.ReadAll(resp.Body)
	if err != nil {
		return err
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return fmt.Errorf("throwawaymail: http %d", resp.StatusCode)
	}
	if out == nil {
		return nil
	}
	return json.Unmarshal(raw, out)
}

func throwawaymailRaw(summary throwawaymailMessageSummary, detail *throwawaymailMessageDetail, recipient string) map[string]any {
	raw := map[string]any{
		"id":           summary.MessageID,
		"messageId":    summary.MessageID,
		"from_address": summary.FromAddress,
		"subject":      summary.Subject,
		"received_at":  summary.ReceivedAt,
		"read":         summary.Read,
		"to":           recipient,
		"size":         summary.Size,
	}
	if summary.FromName != nil {
		raw["fromName"] = *summary.FromName
	}
	if detail != nil {
		if len(detail.ToAddresses) > 0 {
			raw["to"] = detail.ToAddresses[0]
		}
		if detail.Text != nil {
			raw["text"] = *detail.Text
		}
		if detail.HTML != nil {
			raw["html"] = *detail.HTML
		}
	}
	return raw
}

func ThrowawaymailGenerate() (*CreatedMailbox, error) {
	var data throwawaymailMailboxResponse
	if err := throwawaymailJSON(http.MethodPost, throwawaymailBase+"/mailboxes", nil, &data); err != nil {
		return nil, err
	}
	mailboxID := strings.TrimSpace(data.MailboxID)
	email := strings.TrimSpace(data.Address)
	if mailboxID == "" || email == "" || !strings.Contains(email, "@") {
		return nil, fmt.Errorf("throwawaymail: invalid mailbox response")
	}
	return &CreatedMailbox{
		Channel:   "throwawaymail",
		Email:     email,
		Token:     mailboxID,
		ExpiresAt: data.ExpiresAt,
		CreatedAt: data.CreatedAt,
	}, nil
}

func throwawaymailFetchDetail(mailboxID, messageID string) (*throwawaymailMessageDetail, error) {
	u := fmt.Sprintf(
		"%s/mailboxes/%s/messages/%s",
		throwawaymailBase,
		url.PathEscape(mailboxID),
		url.PathEscape(messageID),
	)
	var detail throwawaymailMessageDetail
	if err := throwawaymailJSON(http.MethodGet, u, nil, &detail); err != nil {
		return nil, err
	}
	return &detail, nil
}

func ThrowawaymailGetEmails(mailboxID, email string) ([]NormEmail, error) {
	mailboxID = strings.TrimSpace(mailboxID)
	email = strings.TrimSpace(email)
	if mailboxID == "" {
		return nil, fmt.Errorf("throwawaymail: empty mailbox id")
	}
	if email == "" {
		return nil, fmt.Errorf("throwawaymail: empty email")
	}

	u := fmt.Sprintf("%s/mailboxes/%s/messages", throwawaymailBase, url.PathEscape(mailboxID))
	var rows []throwawaymailMessageSummary
	if err := throwawaymailJSON(http.MethodGet, u, nil, &rows); err != nil {
		return nil, err
	}
	out := make([]NormEmail, 0, len(rows))
	for _, row := range rows {
		var detail *throwawaymailMessageDetail
		if strings.TrimSpace(row.MessageID) != "" {
			d, err := throwawaymailFetchDetail(mailboxID, row.MessageID)
			if err == nil {
				detail = d
			}
		}
		out = append(out, NormalizeMap(throwawaymailRaw(row, detail, email), email))
	}
	return out, nil
}

File: sdk/npm/src/providers/throwawaymail.ts
import { InternalEmailInfo, Email, Channel } from '../types';
import { normalizeEmail } from '../normalize';
import { fetchWithTimeout } from '../retry';

const CHANNEL: Channel = 'throwawaymail';
const API_BASE = 'https://throwawaymail.app/api';

interface MailboxResponse {
  mailbox_id?: string;
  address?: string;
  expires_at?: string;
  created_at?: string;
}

interface MessageSummaryResponse {
  message_id?: string;
  from_address?: string;
  from_name?: string | null;
  subject?: string;
  received_at?: string;
  size?: number;
  read?: boolean;
}

interface MessageDetailResponse extends MessageSummaryResponse {
  to_addresses?: string[];
  cc_addresses?: string[] | null;
  bcc_addresses?: string[] | null;
  reply_to?: string | null;
  text?: string | null;
  html?: string | null;
}

function normalizeRaw(raw: MessageSummaryResponse | MessageDetailResponse, recipient: string): Record<string, unknown> {
  return {
    id: raw.message_id,
    messageId: raw.message_id,
    from_address: raw.from_address,
    fromName: raw.from_name,
    to: Array.isArray((raw as MessageDetailResponse).to_addresses) && (raw as MessageDetailResponse).to_addresses!.length > 0
      ? (raw as MessageDetailResponse).to_addresses![0]
      : recipient,
    subject: raw.subject || '',
    received_at: raw.received_at,
    read: raw.read,
    text: (raw as MessageDetailResponse).text || '',
    html: (raw as MessageDetailResponse).html || '',
    size: raw.size,
  };
}

async function fetchJSON<T>(url: string, init?: RequestInit): Promise<T> {
  const response = await fetchWithTimeout(url, {
    ...init,
    headers: {
      Accept: 'application/json',
      ...(init?.headers || {}),
    },
  });
  if (!response.ok) {
    throw new Error(`throwawaymail http ${response.status}`);
  }
  return response.json() as Promise<T>;
}

async function fetchDetail(mailboxId: string, messageId: string): Promise<MessageDetailResponse | null> {
  const url = `${API_BASE}/mailboxes/${encodeURIComponent(mailboxId)}/messages/${encodeURIComponent(messageId)}`;
  try {
    return await fetchJSON<MessageDetailResponse>(url);
  } catch {
    return null;
  }
}

export async function generateEmail(): Promise<InternalEmailInfo> {
  const data = await fetchJSON<MailboxResponse>(`${API_BASE}/mailboxes`, { method: 'POST' });
  const mailboxId = String(data.mailbox_id || '').trim();
  const email = String(data.address || '').trim();
  if (!mailboxId || !email || !email.includes('@')) {
    throw new Error('throwawaymail: invalid mailbox response');
  }

  return {
    channel: CHANNEL,
    email,
    token: mailboxId,
    expiresAt: data.expires_at,
    createdAt: data.created_at,
  };
}

export async function getEmails(token: string, email: string): Promise<Email[]> {
  const mailboxId = String(token || '').trim();
  const address = String(email || '').trim();
  if (!mailboxId) throw new Error('throwawaymail: empty mailbox id');
  if (!address) throw new Error('throwawaymail: empty email');

  const url = `${API_BASE}/mailboxes/${encodeURIComponent(mailboxId)}/messages`;
  const rows = await fetchJSON<MessageSummaryResponse[]>(url);
  if (!Array.isArray(rows)) return [];

  const emails: Email[] = [];
  for (const row of rows) {
    const messageId = String(row.message_id || '').trim();
    const detail = messageId ? await fetchDetail(mailboxId, messageId) : null;
    emails.push(normalizeEmail(normalizeRaw(detail || row, address), address));
  }
  return emails;
}
